### PR TITLE
Spa (router & REST api) prototype

### PR DESCRIPTION
**Pull request by Neffi42**
- **intial commit**
- **router prototype**
- **put script at the of body part**
- **pure vanilia router up and running**
- **updated Django to version 5**
- **created rest api**
- **api endpoint to fetch pages is up**

### CODE REVIEW

File: django/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM python:3.12.4-slim
+
+WORKDIR /app
+
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED=1
+
```
Comment (antoineverin) [inline django/Dockerfile:7]: Missing `apt update && apt upgrade -y` and `python3 -m pip install --upgrade pip`

File: django/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+COPY conf/requirements.txt requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY conf/entrypoint.sh /entrypoint.sh
+RUN chmod a+x /entrypoint.sh
+
+ENTRYPOINT [ "/entrypoint.sh" ]
```
Comment (antoineverin) [inline django/Dockerfile:14]: Not correct use of `ENTRYPOINT` should use `CMD` instead

File: django/src/rendering/views.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+from django.template import TemplateDoesNotExist
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from rest_framework import status
+from .serializers import PageSerializer
+
+class FetchPageView(APIView):
+    def get(self, request, page_name):
+        try:
+            html_content = render_to_string(f'pages/{page_name}.html')
```
Comment (antoineverin) [inline django/src/rendering/views.py:11]: Why putting templates in a pages folder?

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      - /static:/static
+      - /media:/media
+    env_file:
+      - .env
+
+  nginx:
+    container_name: nginx
+    build: nginx
+    ports:
+      - 80:80
```
Comment (antoineverin) [inline docker-compose.yml:23]: Port 80 could already be in use on some computers, for development, using 8080 should be a better option

File: docker-compose.yml
Comment (antoineverin) [inline docker-compose.yml:1]: Why creating volumes for the /static and /media folders?

File: nginx/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,5 @@
+FROM nginx:1.26.1
+
```
Comment (antoineverin) [inline nginx/Dockerfile:2]: Missing `apt update && apt upgrade -y`

File: django/src/ft_transcendence/urls.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+from django.urls import path, include
+from rest_framework import routers
+from rendering.views import FetchPageView
+
+router = routers.DefaultRouter()
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+    path('api/', include(router.urls)),
+    path('api/fetch-page/<str:page_name>/', FetchPageView.as_view(), name='fetch_page'),
```
Comment (antoineverin) [inline django/src/ft_transcendence/urls.py:27]: May rename uri to `api/pages` only?

File: django/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED=1
+
+COPY conf/requirements.txt requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY conf/entrypoint.sh /entrypoint.sh
+RUN chmod a+x /entrypoint.sh
+
```
Comment (antoineverin) [inline django/Dockerfile:13]: Missing expose and volume declaration for documentation

File: nginx/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+COPY conf/nginx.conf /etc/nginx/conf.d/default.conf
+
```
Comment (antoineverin) [inline nginx/Dockerfile:4]: Missing expose and volume declaration for documentation

**Review by antoineverin — CHANGES_REQUESTED** [review]

File: django/Dockerfile (hunk 1 above)
Comment (Neffi42) [reply django/Dockerfile:7]: why update apt ? we are not using any package from it

File: nginx/Dockerfile (hunk 5 above)
Comment (Neffi42) [reply nginx/Dockerfile:2]: same as for django, there is no point to update just for the sake of it

File: django/src/rendering/views.py (hunk 3 above)
Comment (Neffi42) [reply django/src/rendering/views.py:11]: yes indeed not usefull

File: django/Dockerfile (hunk 2 above)
Comment (Neffi42) [reply django/Dockerfile:14]: okkk

File: nginx/Dockerfile (hunk 8 above)
Comment (Neffi42) [reply nginx/Dockerfile:4]: ok

File: django/Dockerfile (hunk 7 above)
Comment (Neffi42) [reply django/Dockerfile:13]: ok

File: django/src/ft_transcendence/urls.py (hunk 6 above)
Comment (Neffi42) [reply django/src/ft_transcendence/urls.py:27]: ok yes good idea

File: docker-compose.yml
Comment (Neffi42) [reply docker-compose.yml:1]: I missunderstood something, it was stupid

Comment (Neffi42) [reply docker-compose.yml:1]: because if not the staticfiles from Django cannot be passed to nginx

Comment (Neffi42) [reply docker-compose.yml:1]: but we should create tmp folder to stock them

Comment (Neffi42) [reply docker-compose.yml:1]: you can check what I did if you want

Comment (antoineverin) [reply docker-compose.yml:1]: changed to non project directory

**Review by antoineverin — APPROVED** [review]

**Pull request merged**